### ISSUE TITLE
Create unique list of valid RespondentState and RespondentDisposition

### ISSUE THREAD

**Issue by diegoliberman** [instedd/surveda#1975]
Currently, the codebase is full of strings defining states and dispositions from scratch, we should define them as constants only once.

**Comment by ysbaddaden** [comment]
Now that we have a working Ecto 3, I tried the `enum_type` dependency again... just to understand that it just doesn't work when building SQL queries :sob: 

But Ecto 3.5 introduced `Ecto.Enum` which does the job:

1. it replaces strings with atoms, while still persisting the value as a String (hence no change in the database);
2. we already validated the value when persisting a record, and must continue to do so, but we can use `Ecto.Enum.values(Ask.Respondent, :state)` to avoid duplicating the list (nice);
3. `Ecto.Enum` goes the extra mile and also prevents using an invalid atom in queries :tada:
4. It's baked in Ecto: no extra dependency to install :rocket: 

It means that a query with a typo such as `from r in Respondent, where: state == :cmopleted` will now fail with:

```
** (Ecto.QueryError) iex:2: value `:cmopleted` cannot be dumped to type {:parameterized,
 Ecto.Enum, %{on_dump: %{active: "active", cancelled: "cancelled", completed: "completed",
 failed: "failed", pending: "pending", rejected: "rejected"}, on_load: %{"active" => :active,
 "cancelled" => :cancelled, "completed" => :completed, "failed" => :failed, "pending" =>
 :pending, "rejected" => :rejected}, values: [:pending, :active, :completed, :failed,
 :rejected, :cancelled]}}. Or the value is incompatible or it must be interpolated
 (using ^) so it may be cast accordingly in query:

from r0 in Ask.Respondent,
  where: r0.state == :cmopleted,
  select: r0
```

**Comment by ysbaddaden** [comment]
1. Refactoring `Respondent.state` to be an enum was easy. Just a few replacements. :heavy_check_mark: 
2. Refactoring `Respondent.disposition` proves to be much more challenging :feelsgood: 
   
   `Respondent.disposition` is now an Atom (e.g. `:started`) instead of a String (`"started"`) but Surveda assumes that it's a String everywhere, and there isn't just a respondent disposition. There are:

   - :ok: SurveyLogger.disposition (String): mere call to `to_string(disposition)` for now (but may use `Ecto.Enum` later);
   - :ok: RespondentHistory.disposition (String): same.
   - :warning: Each step of a questionnaire can *change the disposition*... and these steps are stored as a big JSON hash in the database, so we can't use `Ecto.Enum` to do the mapping.

We use the step's disposition value _deep_ into Surveda's runtime (:ocean:), for example to trigger disposition changes, or check whether there is a disposition change (Session, Flow, Step, Reply, ...). 

I'm pondering whether we should handle strings in addition to atoms, but that's prone to become nasty, and sounds opposite to this issue's goal. Maybe creating a new `Ecto.Type` extending `Ecto.Type.JSON` that would transform the questionnaire steps' disposition (when present) from String to Atom would be a better idea? Then, we'd only use atoms everywhere.

**Comment by ysbaddaden** [comment]
This seems like working. Ready for review.

**Comment by ysbaddaden** [comment]
Merged into the `develop` branch.

**Comment by diegoliberman** [comment]
The best way to test this is the Sandbox and create surveys that use all of the dispositions, and test related functionality.  We should use the broker that communicates with Verboice and Nuntium to check the logs and verify that the responses make sense.
